### PR TITLE
Replace boost::optional with std::optional

### PR DESCRIPTION
**Pull request by maflcko**
Now that we can use std::optional from the vanilla standard library, drop the third-party boost dependency

### CODE REVIEW

**Comment by laanwj** [comment]
Nice, concept ACK, this is a surprisingly small patch.

**Comment by hebasto** [comment]
Concept ACK.

**Comment by jonatack** [comment]
Concept ACK -- nice

**Comment by jnewbery** [comment]
Concept ACK.

#19806 could be rebased on this to remove the need for additional boost dependencies: https://github.com/bitcoin/bitcoin/pull/19806#discussion_r543721793.

@MarcoFalke do you still consider #19426 a prerequisite for this? (https://github.com/bitcoin/bitcoin/pull/19183#issuecomment-733637916)

**Comment by practicalswift** [comment]
Concept ACK: nice to get rid of old cruft

**Comment by MarcoFalke** [comment]
> @MarcoFalke do you still consider #19426 a prerequisite for this? (#19183 (comment))

I was hoping to get that in first, so that the first commit is easier to review. Though, it is not a strict requirement.

**Comment by MarcoFalke** [comment]
References for the third commit:

https://www.boost.org/doc/libs/1_75_0/libs/optional/doc/html/boost_optional/reference/header__boost_optional_optional_hpp_/detailed_semantics___optional_values.html

https://en.cppreference.com/w/cpp/utility/optional

**Comment by promag** [comment]
Concept ACK.

File: src/optional.h [hunk 1]
```diff
@@ -5,22 +5,16 @@
 #ifndef BITCOIN_OPTIONAL_H
 #define BITCOIN_OPTIONAL_H
 
+#include <optional>
 #include <utility>
 
-#include <boost/optional.hpp>
-
 //! Substitute for C++17 std::optional
+//! DEPRECATED use std::optional in new code.
```
Comment (promag) [inline src/optional.h:12]: To avoid conflicts with other pulls or backports? If so it could mention an acceptable time to remove this alias.

Comment (MarcoFalke) [reply src/optional.h:12]: I'd say when 0.21.1 is out or when 0.22 is about to get branched off, whichever happens first.

Comment (promag) [reply src/optional.h:12]: Sounds good to me.

**Comment by jnewbery** [comment]
> Replace get_ptr() with &value() in UpdatePSBTOutput
> This is safe because a value is always contained in this optional

What do you think about explicitly stating that assumption at the top of the function:

```diff
--- a/src/psbt.cpp
+++ b/src/psbt.cpp
@@ -207,7 +207,11 @@ size_t CountPSBTUnsignedInputs(const PartiallySignedTransaction& psbt) {
 
 void UpdatePSBTOutput(const SigningProvider& provider, PartiallySignedTransaction& psbt, int index)
 {
-    const CTxOut& out = psbt.tx->vout.at(index);
+    // tx must be filled before updating this PSBT.
+    assert(psbt.tx.has_value());
+    CMutableTransaction& tx = psbt.tx.value();
+
+    const CTxOut& out = tx.vout.at(index);
     PSBTOutput& psbt_out = psbt.outputs.at(index);
 
     // Fill a SignatureData with output info
@@ -217,7 +221,7 @@ void UpdatePSBTOutput(const SigningProvider& provider, PartiallySignedTransactio
     // Construct a would-be spend of this output, to update sigdata with.
     // Note that ProduceSignature is used to fill in metadata (not actual signatures),
     // so provider does not need to provide any private keys (it can be a HidingSigningProvider).
-    MutableTransactionSignatureCreator creator(&psbt.tx.value(), /* index */ 0, out.nValue, SIGHASH_ALL);
+    MutableTransactionSignatureCreator creator(&tx, /* index */ 0, out.nValue, SIGHASH_ALL);
```

**Comment by practicalswift** [comment]
cr ACK fa4435e22f78f632a455016ce00a357009aac059: patch looks correct!

File: src/psbt.cpp [hunk 2]
```diff
@@ -207,7 +209,8 @@ size_t CountPSBTUnsignedInputs(const PartiallySignedTransaction& psbt) {
 
 void UpdatePSBTOutput(const SigningProvider& provider, PartiallySignedTransaction& psbt, int index)
 {
-    const CTxOut& out = psbt.tx->vout.at(index);
+    CMutableTransaction& tx = *Assert(psbt.tx);
```
Comment (hebasto) [inline src/psbt.cpp:212]: nit: to be explicit
```suggestion
    CMutableTransaction& tx = *Assert(psbt.tx.has_value());
```
as were [suggested](https://github.com/bitcoin/bitcoin/pull/20671#issuecomment-748046093) by @jnewbery.

**Review by hebasto — APPROVED** [review]
ACK fa4435e22f78f632a455016ce00a357009aac059, I have reviewed the code and it looks OK, I agree it can be merged.

File: src/psbt.cpp (hunk 2 above)
Comment (MarcoFalke) [reply src/psbt.cpp:212]: That wouldn't compile. Also, what is the benefit of it?

Comment (hebasto) [reply src/psbt.cpp:212]: > That wouldn't compile.

Indeed :man_facepalming: 

Sorry for noice.

**Comment by laanwj** [comment]
code review ACK fa4435e22f78f632a455016ce00a357009aac059

**Comment by apoelstra** [comment]
Since this PR many of the functional tests fail with `--valgrind`. This feels like a compiler bug. It doesn't happen with `--enable-debug` and stops happening when I try to add trace code or extra conditonals.

e.g. on 68c7acf6bb7733261865afdcc05ee687ef75489d
```
19:25:55 -bash@camus ~/code/bitcoin/bitcoin/master 68c7acf6bb7$ ./test/functional/interface_bitcoin_cli.py --valgrind
2021-08-03T19:25:55.328000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_h1zfvbq7
2021-08-03T19:26:04.859000Z TestFramework (ERROR): Unexpected exception caught during testing
Traceback (most recent call last):
    <python noise redacted, just saying "stderr was not empty">
AssertionError: Unexpected stderr ==1787451== Thread 19 b-httpworker.0:
==1787451== Conditional jump or move depends on uninitialised value(s)
==1787451==    at 0x5074E0: CWallet::ScanForWalletTransactions(uint256 const&, int, std::optional<int>, WalletRescanReserver const&, bool) (wallet.cpp:1805)
==1787451==    by 0x50809C: CWallet::RescanFromTime(long, WalletRescanReserver const&, bool) (wallet.cpp:1708)
==1787451==    by 0x5581DB: RescanWallet(CWallet&, WalletRescanReserver const&, long, bool) (rpcdump.cpp:85)
==1787451==    by 0x57537E: importprivkey()::{lambda(RPCHelpMan const&, JSONRPCRequest const&)#1}::operator()(RPCHelpMan const&, JSONRPCRequest const&) const [clone .constprop.0] (rpcdump.cpp:188)
==1787451==    by 0x575694: __invoke_impl<UniValue, importprivkey()::<lambda(const RPCHelpMan&, const JSONRPCRequest&)>&, const RPCHelpMan&, const JSONRPCRequest&> (invoke.h:61)
==1787451==    by 0x575694: __invoke_r<UniValue, importprivkey()::<lambda(const RPCHelpMan&, const JSONRPCRequest&)>&, const RPCHelpMan&, const JSONRPCRequest&> (invoke.h:116)
==1787451==    by 0x575694: std::_Function_handler<UniValue (RPCHelpMan const&, JSONRPCRequest const&), importprivkey()::{lambda(RPCHelpMan const&, JSONRPCRequest const&)#1}>::_M_invoke(std::_Any_data const&, RPCHelpMan const&, JSONRPCRequest const&) (std_function.h:292)
==1787451==    by 0x2D4A9E: operator() (std_function.h:560)
==1787451==    by 0x2D4A9E: HandleRequest (util.h:343)
==1787451==    by 0x2D4A9E: CRPCCommand::CRPCCommand(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, RPCHelpMan (*)(), std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >)::{lambda(JSONRPCRequest const&, UniValue&, bool)#1}::operator()(JSONRPCRequest const&, UniValue&, bool) const [clone .isra.0] (server.h:110)
==1787451==    by 0x43C473: operator() (std_function.h:560)
==1787451==    by 0x43C473: wallet::(anonymous namespace)::WalletClientImpl::registerRpcs()::{lambda(JSONRPCRequest const&, UniValue&, bool)#1}::operator()(JSONRPCRequest const&, UniValue&, bool) const [clone .isra.0] (interfaces.cpp:517)
==1787451==    by 0x251DE4: operator() (std_function.h:560)
==1787451==    by 0x251DE4: operator() (interfaces.cpp:382)
==1787451==    by 0x251DE4: __invoke_impl<bool, node::(anonymous namespace)::RpcHandlerImpl::RpcHandlerImpl(const CRPCCommand&)::<lambda(const JSONRPCRequest&, UniValue&, bool)>&, const JSONRPCRequest&, UniValue&, bool> (invoke.h:61)
==1787451==    by 0x251DE4: __invoke_r<bool, node::(anonymous namespace)::RpcHandlerImpl::RpcHandlerImpl(const CRPCCommand&)::<lambda(const JSONRPCRequest&, UniValue&, bool)>&, const JSONRPCRequest&, UniValue&, bool> (invoke.h:114)
==1787451==    by 0x251DE4: std::_Function_handler<bool (JSONRPCRequest const&, UniValue&, bool), node::(anonymous namespace)::RpcHandlerImpl::RpcHandlerImpl(CRPCCommand const&)::{lambda(JSONRPCRequest const&, UniValue&, bool)#1}>::_M_invoke(std::_Any_data const&, JSONRPCRequest const&, UniValue&, bool&&) (std_function.h:291)
==1787451==    by 0x352364: operator() (std_function.h:560)
==1787451==    by 0x352364: ExecuteCommand (server.cpp:466)
==1787451==    by 0x352364: CRPCTable::execute(JSONRPCRequest const&) const (server.cpp:449)
==1787451==    by 0x402AB9: HTTPReq_JSONRPC(util::Ref const&, HTTPRequest*) (httprpc.cpp:201)
==1787451==    by 0x40ED8C: operator() (std_function.h:560)
==1787451==    by 0x40ED8C: operator() (httpserver.cpp:49)
==1787451==    by 0x40ED8C: WorkQueue<HTTPClosure>::Run() (httpserver.cpp:108)
==1787451==    by 0x40A7C7: HTTPWorkQueueRun(WorkQueue<HTTPClosure>*, int) (httpserver.cpp:336)
==1787451== 
{
   <insert_a_suppression_name_here>
   Memcheck:Cond
   fun:_ZN7CWallet25ScanForWalletTransactionsERK7uint256iSt8optionalIiERK20WalletRescanReserverb
   fun:_ZN7CWallet14RescanFromTimeElRK20WalletRescanReserverb
   fun:_ZL12RescanWalletR7CWalletRK20WalletRescanReserverlb
   fun:_ZZ13importprivkeyvENKUlRK10RPCHelpManRK14JSONRPCRequestE_clES1_S4_.constprop.0
   fun:__invoke_impl<UniValue, importprivkey()::<lambda(const RPCHelpMan&, const JSONRPCRequest&)>&, const RPCHelpMan&, const JSONRPCRequest&>
   fun:__invoke_r<UniValue, importprivkey()::<lambda(const RPCHelpMan&, const JSONRPCRequest&)>&, const RPCHelpMan&, const JSONRPCRequest&>
   fun:_ZNSt17_Function_handlerIF8UniValueRK10RPCHelpManRK14JSONRPCRequestEZ13importprivkeyvEUlS3_S6_E_E9_M_invokeERKSt9_Any_dataS3_S6_
   fun:operator()
   fun:HandleRequest
   fun:_ZZN11CRPCCommandC4ENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES5_PF10RPCHelpManvESt6vectorIS5_SaIS5_EEENKUlRK14JSONRPCRequestR8UniValuebE_clESE_SG_b.isra.0
   fun:operator()
   fun:_ZZN6wallet12_GLOBAL__N_116WalletClientImpl12registerRpcsEvENKUlRK14JSONRPCRequestR8UniValuebE_clES4_S6_b.isra.0
   fun:operator()
   fun:operator()
   fun:__invoke_impl<bool, node::(anonymous namespace)::RpcHandlerImpl::RpcHandlerImpl(const CRPCCommand&)::<lambda(const JSONRPCRequest&, UniValue&, bool)>&, const JSONRPCRequest&, UniValue&, bool>
   fun:__invoke_r<bool, node::(anonymous namespace)::RpcHandlerImpl::RpcHandlerImpl(const CRPCCommand&)::<lambda(const JSONRPCRequest&, UniValue&, bool)>&, const JSONRPCRequest&, UniValue&, bool>
   fun:_ZNSt17_Function_handlerIFbRK14JSONRPCRequestR8UniValuebEZN4node12_GLOBAL__N_114RpcHandlerImplC4ERK11CRPCCommandEUlS2_S4_bE_E9_M_invokeERKSt9_Any_dataS2_S4_Ob
   fun:operator()
   fun:ExecuteCommand
   fun:_ZNK9CRPCTable7executeERK14JSONRPCRequest
   fun:_ZL15HTTPReq_JSONRPCRKN4util3RefEP11HTTPRequest
   fun:operator()
   fun:operator()
   fun:_ZN9WorkQueueI11HTTPClosureE3RunEv
   fun:_ZL16HTTPWorkQueueRunP9WorkQueueI11HTTPClosureEi
}
==1787451== 
==1787451== Exit program on first error (--exit-on-first-error=yes) != 
[node 0] Cleaning up leftover process

```

**Comment by apoelstra** [comment]
I think it is https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80635

**Comment by MarcoFalke** [comment]
See also https://github.com/bitcoin/bitcoin/pull/18785, an unrelated but similar issue

**Pull request merged**